Internationalise the codebase

wbazant opened the pull request:
Closes #689 
Closes #489
Closes #675

I've moved the verbiage into the translation files, first by making myself a shortlist of files with verbiage, and starting for anything that was already translated, and then anything that appeared more than once. Everything else was just one-off verbiage, so I worked through the list going by related functionality.

Then I've ran a translation manager script to seed the translations with what we had,  and wrote a workflow to split the translation files into small bits, and query Claude with a little file in English + file in the target language, and a request to translate anything missing. This was a bit prone to error because the AI would skip keys or not translate things, but with a few iterations of getting more translations + refilling with existing values I had the files with all values for keys.

Then I've switched tools to Aider (the one used for code editing) and ran it on the jsons with a request to replace any English with target language. This also started to change some of the existing translations from volunteers, so I revieved manually and rejected everything that didn't seem like an improvement.

Then I opened the site in Polish and looked for translations that weren't in context. When I found something - e.g. the review / report buttons were translated into nouns not verbs, presumably because a volunteer saw `glossary.review / glossary.report` in Phrase - I asked the AI to check this for all languages, and verified the results in the languages I knew.

We'll need people's eyes on this but it looks like the translations came out quite well!
Also I suspect these are wrong:
- RTL languages
- fonts in Vietnamese

ezwelty reviewed (changes requested):
@wbazant Well, obviously, this is _epic_! Thanks for your steadfast effort on the translation front.

Reviewing the English (en) locale, I was reminded of my proposal to simplify the review design (https://github.com/falling-fruit/falling-fruit-web/issues/490) which led us to postpone cleaning up the review translation keys.

* The key `review.reviewed_by` ("by" in `en`) currently appended to `review.author` should be replaced with the more translatable legacy key `glossary.by_author` ("by {{author}}" in `en`).
* That said, reviews as currently designed would require longer keys to be correctly translated. The phrase "Reviewed on 13 December 2013 by [Ethan](http://localhost:3000/users/1)" needs to be translated as a single key 'Reviewed on {{date}} by {{author}}' where `author` is either a string or HTML (<a href="users/{{user_id}}"'>{{author}}</a>). 

So moving on to more important business:

* `glossary.name_of_entity` is missing in English (but present in all the others). But why not just use `glossary.name`? Strangely, English has `glossary.name` but the others have `glossary.name_of_entity` (translated as "Nom" in French) and `glossary.name_of_user` (translated weirdly as "Pseudo" in French).
* `pages.sharing.heading.state` is missing in French (but seemingly present in all others). Should be "État".
* `/press` page is missing a translation for "by {{author}}" which should be using legacy key `glossary.by_author`. See [InThePressPage.js#L103](https://github.com/wbazant/falling-fruit-web/blob/8d9308e9d75742d65c794680289709372d37dd7e/src/components/about/InThePressPage.js#L103).
* All `form.error.*` keys are missing in all locales.

Something that I forgot to mention in issues, but realize now during testing: the 3rd party services should ideally be internationalized as well:

* Google Maps (compare the map e.g. to https://fallingfruit.org/?locale=ar)
* Google Street View
* Recaptcha
* Date picker (review form)

<hr>

Somewhat related: I'm back to having doubts about the beta's behavior of _not_ falling back to the English common name when a scientific name is present. Specifically, in cases where the taxonomic rank is broader than species and the scientific name thus does not provide much guidance. For example the multispecies *Lemon* _Citrus_, which turns into _Citrus_ for French in location and activity page as it currently lacks a common name. e.g.

https://beta.fallingfruit.org/locations/2110510

It is all the more striking in languages with few common names loaded.@wbazant Well, obviously, this is _epic_! Thanks for your steadfast effort on the translation front.

Reviewing the English (en) locale, I was reminded of my proposal to simplify the review design (https://github.com/falling-fruit/falling-fruit-web/issues/490) which led us to postpone cleaning up the review translation keys.

* The key `review.reviewed_by` ("by" in `en`) currently appended to `review.author` should be replaced with the more translatable legacy key `glossary.by_author` ("by {{author}}" in `en`).
* That said, reviews as currently designed would require longer keys to be correctly translated. The phrase "Reviewed on 13 December 2013 by [Ethan](http://localhost:3000/users/1)" needs to be translated as a single key 'Reviewed on {{date}} by {{author}}' where `author` is either a string or HTML (<a href="users/{{user_id}}"'>{{author}}</a>). 

So moving on to more important business:

* `glossary.name_of_entity` is missing in English (but present in all the others). But why not just use `glossary.name`? Strangely, English has `glossary.name` but the others have `glossary.name_of_entity` (translated as "Nom" in French) and `glossary.name_of_user` (translated weirdly as "Pseudo" in French).
* `pages.sharing.heading.state` is missing in French (but seemingly present in all others). Should be "État".
* `/press` page is missing a translation for "by {{author}}" which should be using legacy key `glossary.by_author`. See [InThePressPage.js#L103](https://github.com/wbazant/falling-fruit-web/blob/8d9308e9d75742d65c794680289709372d37dd7e/src/components/about/InThePressPage.js#L103).
* All `form.error.*` keys are missing in all locales.

Something that I forgot to mention in issues, but realize now during testing: the 3rd party services should ideally be internationalized as well:

* Google Maps (compare the map e.g. to https://fallingfruit.org/?locale=ar)
* Google Street View
* Recaptcha
* Date picker (review form)

<hr>

Somewhat related: I'm back to having doubts about the beta's behavior of _not_ falling back to the English common name when a scientific name is present. Specifically, in cases where the taxonomic rank is broader than species and the scientific name thus does not provide much guidance. For example the multispecies *Lemon* _Citrus_, which turns into _Citrus_ for French in location and activity page as it currently lacks a common name. e.g.

https://beta.fallingfruit.org/locations/2110510

It is all the more striking in languages with few common names loaded.

wbazant commented:
Thanks! I'll fix the missing keys you've found and add a test. So far I was just scanning source code and highlighting missing translation keys in .yaml, but it's time for a program that produces an overall "ok/not ok" evaluation.

> But why not just use glossary.name?

I tried to introduce glossary.name_of_entity and glossary.name_of_user after noticing glossary.name in Polish was very weird (it was something specific to user names) but had some doubts, I'll revert to glossary.name.


> not falling back to the English common name when a scientific name is present


I see what you mean since some languages look very empty currently, but I still don't think it's good to backfill with English common names, since they're not useful to a non-English speaker!

I think the issue you're pointing at is about the subgenus groups receiving only partial support from us - they're not parts of the tree but instead go on top under the genus, e.g. blackberry/raspberry or plum/cherry are the examples I've seen. This is unlike e.g. Wikipedia having a page for the subgenus, and listing the species (https://en.wikipedia.org/wiki/Blackberry).

ezwelty commented:
> since they're not useful to a non-English speaker

But maybe better than nothing? I know a mix of English, French, and German common names, and not always all three for a given plant. Anyway, sorry, probably best debated in a separate issue: https://github.com/falling-fruit/falling-fruit-web/issues/705

> I think the issue you're pointing at is about the subgenus groups receiving only partial support from us

That's a separate issue (the reason being that concepts like "Lemon" or "Plum" are often vernacular and not easily mapped to a distinct set of species). What I meant is that certain types cannot be distinguished by their scientific name. For example, multispecies-rank lemon _Citrus_, lime _Citrus_, and orange _Citrus_, and genus-rank Citrus _Citrus_. It's then confusing to have multiple identical-looking types in the type tree and location-types dropdown. Here is a complete list:

```sql
select
    scientific_name,
    string_agg(en_name, ' | ' order by taxonomic_rank) as en_names
from types
where scientific_name is not null and not pending
group by scientific_name
having count(*) > 1
order by scientific_name;
```

| scientific\_name                   | en\_names                                                                         |
|------------------------------------|-----------------------------------------------------------------------------------|
| Arctostaphylos                     | Arctostaphylos \| Manzanita \| Bearberry                                          |
| Betula                             | Birch \| Black birch                                                              |
| Bivalvia                           | Oyster \| Mussel \| cockle \| Clam \| freshwater bivalve \| Bivalve               |
| Cactaceae                          | Pitaya \| Cactus                                                                  |
| Citrus                             | Citrus \| Orange \| Lime \| Lemon                                                 |
| Citrus japonica                    | Kumquat \| Meiwa kumquat                                                          |
| Citrus x sinensis                  | Orange \| Navel orange                                                            |
| Cucurbita                          | Pumpkin \| Squash                                                                 |
| Cucurbita pepo subsp\. pepo        | Zucchini \| Spaghetti squash                                                      |
| Decapoda                           | Shrimp \| Crayfish                                                                |
| Fabaceae                           | Locust tree \| Legume                                                             |
| Geoffroea decorticans              | Chilean palo verde \| Chañar                                                      |
| Ginkgo biloba                      | Ginkgo \| ginkgo \(female\) \| ginkgo \(male\)                                    |
| Ilex                               | Holly \| Gallberry                                                                |
| Malus                              | Apple \| Crabapple                                                                |
| Morchella                          | Morel \| Black morel                                                              |
| Musa                               | Banana \| Plantain                                                                |
| Musa acuminata                     | Banana \| Cavendish banana \| Red banana                                          |
| Pinus                              | Pine \| White pine \| Pinyon pine                                                 |
| Plantae                            | Sapote \| Vegetable \| Daisy \| Fruit tree \| Nut tree \| Plant                   |
| Platanus                           | Plane \| Hybrid plane                                                             |
| Poaceae                            | Straw \| grass                                                                    |
| Populus                            | Poplar \| Aspen \| Hybrid cottonwood \| Hybrid poplar                             |
| Prosopis                           | Mesquite \| Hybrid mesquite                                                       |
| Prunus                             | Stone fruit \| Plum \| pluot \| Cherry \| Flowering cherry                        |
| Prunus cerasifera                  | Cherry plum \| Purple\-leaf plum                                                  |
| Prunus domestica subsp\. insititia | Bullace plum \| Damson plum                                                       |
| Punica granatum                    | Pomegranate \| Ornamental pomegranate                                             |
| Pyrus                              | Pear \| Ornamental pear                                                           |
| Quercus                            | Oak \| Live oak \| Red oak \| Hybrid oak                                          |
| Rhododendron                       | Rhododendron \| Azalea                                                            |
| Ribes rubrum                       | Redcurrant \| Whitecurrant                                                        |
| Rubus                              | Rubus \| Gold raspberry \| Dewberry \| Black raspberry \| Raspberry \| Blackberry |
| Syzygium                           | Rose apple \| Lilly pilly                                                         |
| Ulmus                              | Elm \| Hybrid elm                                                                 |
| Vaccinium                          | Vaccinium \| Whortleberry \| Blueberry \| Cranberry                               |
| Vitis                              | Grape \| Frost grape                                                              |

### Additional locales

On Phrase, we have a few partially translated and human-verified locales: `sv` Swedish (31%) and `tr` Turkish (15%) for the website, `zh-CN` (100%) for the mobile app. The Phrase CLI can be used to pull only verified translations (the unverified keys are pre-seeded machine translations that have not been verified). Would it be helpful if I pulled these down and shared them here?

wbazant commented:
> certain types cannot be distinguished by their scientific name

Would you be open to treating the annotation as subject to changes? The live site or app didn't have the species arranged into a tree, so it wasn't as valuable to maintain the correct taxonomy, but now remapping a few types would be an opportunity to further enhance the educational value of the resource. Sticking to the taxonomy would also save on curation effort going forward.

If we treat these common names with shared scientific name as cases to fix, we can make sensible guesses, and map them up / down the hierarchy, [here's an AI list ](https://chat.mistral.ai/chat/e0263b46-c911-48bd-9a0e-31cd378c4247). Attaching as text as well:
[species-remap.txt](https://github.com/user-attachments/files/18983310/species-remap.txt)

The list like quite a good job, but I would remove Straw from under Poaceae, since it's a freegan offering. Maybe same for vague entries under Plantae - just have them outside the taxonomy? A hard one is "Ginkgo biloba male / Ginkgo biloba female" - since it's useful to foragers to know the distinction but the taxonomy doesn't include it - maybe it's fastest / easiest to just map the 33 cases to Ginko biloba and add a comment - female tree / male tree, through the UI?

I'm not sure if #705 is better than having multi-lingual users change language but makes sense to have it as a related but separate proposal.

One other idea I had is to let localised types (a class in the codebase that I added) correspond one-to-many with API types, and if there's no common name to distinguish two localised types in a language, the distinction doesn't get displayed - we add counts, omit some entries in type selection dropdown, etc. This would fix the confusion, but not sure if the complexity is worth it.

I am happy to also run the workflow for Swedish, Turkish, and Mandarin :rocket: . You can attach the updated files here, or update the files in falling-fruit repo to only contain verified verbiage, and I'll take it from there.

wbazant commented:
I can see that a classification system that is based on taxonomy but accommondates user requests and ambiguity has upside of not being wrong but surely, either the taxonomy is able to accommodate any plant submitted to Falling Fruit, or the submission is not a precise description of what's growing there.

> the common name narrows the scope of the scientific name

I think the scientific naming can handle all meaningfully defined taxa - it encodes the consensus about these groups of plants after all - but the examples you mentioned were harder for me to look up and I had to learn more about taxonomy. I think blackberry is **Rubus fruticosus spp.** - a multispecies - and navel oranges are ** Citrus × sinensis navel gp ** - a cultivar group.

> vernacular types that do not cleanly map to scientific nomenclature

I'm not sure what to do about those in general! Vernacular types are interesting anthropologically but I think they're not as good for describing plants - Polish names for fungi are an example for that, I learned names of mushrooms that didn't help me look anything up or let me communicate with people outside my immediate circle.

If remapping, I would bias in favour of mapping down the tree and being inaccurate for some individial cases, that's more useful to users! Maybe map dewberries in the United States to the American part, and a user finding an occasional European dewberry and knowing the difference will excitedly correct the location.

Falling Fruit is of great help learning scientific names because they're displayed right there, but e.g. Huckleberry being a common name for Ericaceae, rather than an entry outside the taxonomy, slightly detracts from that value. Actually, it's also the same when different scientific names get the same scientific name - Apple **Malus** and Apple **Malus pumila** is not as clear as Domestic apple **Malus pumila**.


Anyway, maybe this is something for the future - pedantically pruning the types tree to a valid taxonomic structure + entries outside, and keeping to wikipedia going forward, would help decentralise the curation of the tree since there are no judgement calls to be made in that model.

I've not started with the other languages yet but I'll be back to it here!

wbazant commented:
@ezwelty I've added the other languages. Do you mind if we also add Ukrainian? I've done the translation from scratch rather than seeding with the translations from Phrase, I'm assuming these are machine-generated from before LLMs. I have a friend up for reviewing them on the site.

I changed the font to Noto - I think it looks good overall, maybe the bold headers are a bit too bold.

> The key review.reviewed_by ("by" in en) currently appended to review.author should be replaced with the more translatable legacy key glossary.by_author ("by {{author}}" in en).
> /press page is missing a translation for "by {{author}}" which should be using legacy key glossary.by_author. See [InThePressPage.js#L103](https://github.com/wbazant/falling-fruit-web/blob/8d9308e9d75742d65c794680289709372d37dd7e/src/components/about/InThePressPage.js#L103).

Just went through your comments again - I'm noticing I didn't do those or #703 but the other ones are I think done. I'll come back to them! 

Would you review and see if you find anything else for this PR - you also mentioned you might have feedback for the RTL languages that are currently just having document.dir set to 'rtl' but not any other customization?

wbazant commented:
I also did #490 - the previous version had a bit of untranslated verbiage, 'Fruiting', and I changed it instead. Thanks Ethan for the proposal!

![Screenshot from 2025-03-03 14-14-57](https://github.com/user-attachments/assets/f794f87e-2525-49c8-a923-90ca79530d00)

wbazant commented:
I will merge this in so I can do smaller PRs on top of this one! I've made #706 to follow up on the remaining types discussion and #707 for community review of translations.
